src/lib/homestore_backend/hs_blob_manager.cpp

```diff
@@ -1,17 +1,271 @@
 #include "hs_homeobject.hpp"
+#include "replication_message.hpp"
+#include "replication_state_machine.hpp"
+#include "lib/homeobject_impl.hpp"
+
+SISL_LOGGING_DECL(blobmgr)
 
 namespace homeobject {
 
-BlobManager::Result< blob_id_t > HSHomeObject::_put_blob(ShardInfo const&, Blob&&) {
-    return folly::makeUnexpected(BlobError::UNKNOWN);
+BlobManager::AsyncResult< blob_id_t > HSHomeObject::_put_blob(ShardInfo const& shard, Blob&& blob) {
+    auto& pg_id = shard.placement_group;
+    shared< homestore::ReplDev > repl_dev;
+    {
+        std::shared_lock lock_guard(_pg_lock);
+        auto iter = _pg_map.find(pg_id);
+        if (iter == _pg_map.end()) {
+            LOGERROR("failed to put blob with non-exist pg [{}]", pg_id);
+            return folly::makeUnexpected(BlobError::UNKNOWN_PG);
+        }
+        repl_dev = static_cast< HS_PG* >(iter->second.get())->repl_dev_;
+    }
+
+    RELEASE_ASSERT(repl_dev != nullptr, "Repl dev instance null");
+
+    const uint32_t needed_size = sizeof(ReplicationMessageHeader);
+    auto req = repl_result_ctx< BlobManager::Result< BlobInfo > >::make(needed_size, 512);
+
+    uint8_t* raw_ptr = req->hdr_buf_.bytes;
+    ReplicationMessageHeader* header = new (raw_ptr) ReplicationMessageHeader();
```

Comment:
in the repl_result_ctx struct,  a ReplicationMessageHeader is already been defined, can we reuse that to avoid re-create a new one? if we do that, req->hdr_buf_ can be used to hold the(blob_header, body and user_key),the benefits is when req is released, buf will be automatically released too.

Reply:
Yeah we could do the first one. But what will we pass in needed_size in that case ? 0 to malloc inside blob ? . I dont prefer option 2 because it will difficult to understand the code. we are anyway trying to reuse the body and user key we got from SM if its aligned and keeping in sg vec.